README.md
# photo

[![pub package](https://img.shields.io/pub/v/photo.svg)](https://pub.dartlang.org/packages/photo) [![pub package](https://img.shields.io/pub/v/photo?include_prereleases)](https://pub.dartlang.org/packages/photo)
[![GitHub](https://img.shields.io/github/license/CaiJingLong/flutter_photo.svg?style=flat-square)](https://github.com/CaiJingLong/flutter_photo)
[![GitHub stars](https://img.shields.io/github/stars/CaiJingLong/flutter_photo.svg?style=social&label=Stars)](https://github.com/CaiJingLong/flutter_photo)

image picker, multi picker

support ios icloud

support video

use flutter as ui

if you want to build custom ui, you just need api to make custom ui. to use [photo_manager](https://github.com/CaiJingLong/flutter_photo_manager) or fork the library to custom ui.

## Screenshot

![image](https://github.com/CaiJingLong/some_asset/blob/master/image_picker1.gif)

## install

latest version : [![pub package](https://img.shields.io/pub/v/photo.svg)](https://pub.dartlang.org/packages/photo) [![pub package](https://img.shields.io/pub/v/photo?include_prereleases)](https://pub.dartlang.org/packages/photo)

```yaml
dependencies:
  photo: $latest_version
```

## Import

```dart
import 'package:photo/photo.dart';
import 'package:photo_manager/photo_manager.dart';
```

## Usage

### Simple use

```dart
void pickAssets() async {
    List<AssetEntity> assetList = await PhotoPicker.pickAsset(context: context);
    /// Use assetList to do something.
}
```

### More option

The context is required, other params is optional.

```dart
void pickAsset() async {
    List<AssetEntity> imgList = await PhotoPicker.pickAsset(
      context: context,
      // BuildContext requied

      /// The following are optional parameters.
      themeColor: Colors.green,
      // the title color and bottom color
      padding: 1.0,
      // item padding
      dividerColor: Colors.grey,
      // divider color
      disableColor: Colors.grey.shade300,
      // the check box disable color
      itemRadio: 0.88,
      // the content item radio
      maxSelected: 8,
      // max picker image count
      provider: I18nProvider.chinese,
      // i18n provider ,default is chinese. , you can custom I18nProvider or use ENProvider()
      rowCount: 5,
      // item row count
      textColor: Colors.white,
      // text color
      thumbSize: 150,
      // preview thumb size , default is 64
      sortDelegate: SortDelegate.common,
      // default is common ,or you make custom delegate to sort your gallery
      checkBoxBuilderDelegate: DefaultCheckBoxBuilderDelegate(
        activeColor: Colors.white,
        unselectedColor: Colors.white,
        checkColor: Colors.blue,
      ), // default is DefaultCheckBoxBuilderDelegate ,or you make custom delegate to create checkbox

      loadingDelegate:
          this, // if you want to build custom loading widget,extends LoadingDelegate [see example/lib/main.dart]

      badgeDelegate: const DefaultBadgeDelegate(), /// or custom class extends [BadgeDelegate]

      pickType: type, // all/image/video

      List<AssetPathEntity> photoPathList, /// when [photoPathList] is not null , [pickType] invalid .

      PhotoPreviousBuilder previousBuilder,
      // custom widget before items, and size should be 64.0 . eg: “camera” button at first.
    );
```

### About photoPathList params

You can use [photo_manager] package to get `List<AssetPathEntity>` and handle or cache.

This parameter is then passed into the `pickAsset` method, where the incoming photoList is rendered instead of the data in the album.

## Whole example

You can see [github](https://github.com/caijinglong/flutter_photo/blob/master/example/) [main.dart](https://github.com/caijinglong/flutter_photo/blob/master/example/lib/main.dart)

## About android

### Migrate to androidX

See the [gitbook](https://caijinglong.gitbooks.io/migrate-flutter-to-androidx/content/)

### Glide

Android native use glide to create image thumb bytes, version is 4.8.0.

If your other android library use the library, and version is not same, then you need edit your android project's build.gradle.

```groovy
rootProject.allprojects {

    subprojects {
        project.configurations.all {
            resolutionStrategy.eachDependency { details ->
                if (details.requested.group == 'com.github.bumptech.glide'
                        && details.requested.name.contains('glide')) {
                    details.useVersion "4.8.0"
                }
            }
        }
    }

}
```

if you use the proguard

see the [github](https://github.com/bumptech/glide#proguard)

## About ios

Because the album is a privacy privilege, you need user permission to access it. You must to modify the `Info.plist` file in Runner project.

like next

```plist
<key>NSPhotoLibraryUsageDescription</key>
<string>App need your agree, can visit your album</string>
```

xcode like image
![in xcode](https://github.com/CaiJingLong/some_asset/blob/master/flutter_photo2.png)

### Build error

if you build error like include of non-modular header inside framework module, see [#10](https://github.com/CaiJingLong/flutter_photo/issues/10) or [so](https://stackoverflow.com/questions/27776497/include-of-non-modular-header-inside-framework-module)

## Thanks

Part of the Android code comes from [debuggerx01](https://github.com/debuggerx01).

## Donate

If my code helps you, and you're willing to buy me a cup of coffee.

you can use [paypal](https://paypal.me/caijinglong)

or scan my alipay

<img width="200px"  src="https://raw.githubusercontent.com/CaiJingLong/asset_for_picgo/master/20181228213953.png"/>


CHANGELOG.md
# CHANGELOG

## 0.5.0

Rollback to 0.5.1 photo_manager

## 0.5.0-dev.3

Fix:

video type error.

Feature:

Clear memory cache.

## 0.5.0-dev.2

- Rollback photo_manager

Fix:

- no selected text

## 0.4.8

- Rollback photo_manager

Fix:

- pickedAssetList error

## 0.4.7

- Rollback photo_manager.
- Add `pickedAssetList` for `pickAsset`.

## 0.4.5+1

Fix:

- `photoPathList` of `PhotoPicker.pickAsset`.

## 0.4.5

Rollback `photo_manager` to `0.4.5`.

## 0.4.4

Rollback `photo_manager` to `0.4.4`.

To fix some error.

## 0.4.3+1

Fix a error for build.

## 0.4.3

Rollback `photo_manager` to `0.4.3`.

## 0.4.2+1

Rollback `photo_manager` to `0.4.2`.

## 0.4.2

## 0.4.1+1

Rollback `photo_manager` to `0.4.1`.

## 0.4.1

**The version will build fail.**

Rollback `photo_manager` to `0.4.1`.

## 0.4.0

**The version will build fail.**

Rollback `photo_manager` to `0.4.0`.

- support androidQ.

## 0.3.4+1

fix:

- title and left icon color.
- checkColor in preview can be edit.

## 0.3.4

Rollback `photo_manager` to `0.3.4`

## 0.3.3

Rollback `photo_manager` to `0.3.3`

## 0.3.2

Rollback `photo_manager` to `0.3.2`

## 0.3.1

`photo_manager` to `0.3.1`

If the selected image is deleted by other applications, the selected status in the selector will be updated correctly.

## 0.3.0 add params photoList

**Breaking change**. Migrate from the deprecated original Android Support Library to AndroidX. This shouldn't result in any functional changes, but it requires any Android apps using this plugin to also migrate if they're using the original support library.

add:

- When the album changes, it refreshes in real time.

fix:

- Video duration badge time problem.
- Images sort by create time.

## 0.2.0

**break change**
support pick only image or video

## 0.1.11

fix pubspec version

## 0.1.10 fix bug

fix a error widget bug.

## 0.1.9

fix all assets i18n provider

## 0.1.8

Fixed crash bug when the number of photos or videos was zero.

## 0.1.7

add a badge delegate for asset

## 0.1.6 Rollback photo_manager version

sort asset by date

## 0.1.5 Rollback photo_manager version

## 0.1.4 fix thumb is null bug

fix thumb bug.

## 0.1.3 support ios icloud asset

## 0.1.2 fix bug

fix all path hasVideo property bug

## 0.1.1 fix bug and add params

add loadingDelegate

## 0.1.0 support video

API incompatibility

ImageXXX rename AssetXXX

## 0.0.8 fix bug

DefaultCheckBoxBuilderDelegate params checkColor not valid bug

## 0.0.7 fix bug

fix dividerColor not valid bug

## 0.0.6 add checkbox delegate

users can use CheckBoxDelegate to custom preview right bottom widget

## 0.0.5 add a params

add the sort delegate to help user sort gallery

Optimized LruCache

add a loading refresh indicator in the gallery

## 0.0.4 fix #1

fix request other permission will crash bug

depo photo_manager 0.0.3

## 0.0.3 add the thumb size to option

add a params for pick image , thumb size

## 0.0.2 fix bug

preview sure button bug
preview bottom safeArea

## 0.0.1 first version

image picker


example/lib/main.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:oktoast/oktoast.dart';
import 'package:photo/photo.dart';
import 'package:photo_manager/photo_manager.dart';

import './preview.dart';
import 'icon_text_button.dart';
import 'picked_example.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return OKToast(
      child: MaterialApp(
        title: 'Pick Image Demo',
        theme: ThemeData(
          primarySwatch: Colors.lime,
        ),
        home: MyHomePage(title: 'Pick Image Demo'),
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> with LoadingDelegate {
  String currentSelected = "";

  @override
  Widget buildBigImageLoading(
      BuildContext context, AssetEntity entity, Color themeColor) {
    return Center(
      child: Container(
        width: 50.0,
        height: 50.0,
        child: CupertinoActivityIndicator(
          radius: 25.0,
        ),
      ),
    );
  }

  @override
  Widget buildPreviewLoading(
      BuildContext context, AssetEntity entity, Color themeColor) {
    return Center(
      child: Container(
        width: 50.0,
        height: 50.0,
        child: CupertinoActivityIndicator(
          radius: 25.0,
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
        actions: <Widget>[
          FlatButton(
            child: Icon(Icons.image),
            onPressed: _testPhotoListParams,
          ),
        ],
      ),
      body: Container(
        child: SingleChildScrollView(
          child: Column(
            children: <Widget>[
              IconTextButton(
                icon: Icons.photo,
                text: "photo",
                onTap: () => _pickAsset(PickType.onlyImage),
              ),
              IconTextButton(
                icon: Icons.videocam,
                text: "video",
                onTap: () => _pickAsset(PickType.onlyVideo),
              ),
              IconTextButton(
                icon: Icons.album,
                text: "all",
                onTap: () => _pickAsset(PickType.all),
              ),
              IconTextButton(
                icon: CupertinoIcons.reply_all,
                text: "Picked asset example.",
                onTap: () => routePage(PickedExample()),
              ),
            ],
          ),
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () => _pickAsset(PickType.all),
        tooltip: 'pickImage',
        child: Icon(Icons.add),
      ),
    );
  }

  void _testPhotoListParams() async {
    var assetPathList =
        await PhotoManager.getAssetPathList(type: RequestType.image);
    _pickAsset(PickType.all, pathList: assetPathList);
  }

  void _pickAsset(PickType type, {List<AssetPathEntity> pathList}) async {
    /// context is required, other params is optional.
    /// context is required, other params is optional.
    /// context is required, other params is optional.

    PhotoPicker.clearThumbMemoryCache();

    List<AssetEntity> imgList = await PhotoPicker.pickAsset(
      // BuildContext required
      context: context,

      /// The following are optional parameters.
      themeColor: Colors.green,
      // the title color and bottom color

      textColor: Colors.white,
      // text color
      padding: 1.0,
      // item padding
      dividerColor: Colors.grey,
      // divider color
      disableColor: Colors.grey.shade300,
      // the check box disable color
      itemRadio: 0.88,
      // the content item radio
      maxSelected: 8,
      // max picker image count
      // provider: I18nProvider.english,
      provider: I18nProvider.chinese,
      // i18n provider ,default is chinese. , you can custom I18nProvider or use ENProvider()
      rowCount: 3,
      // item row count

      thumbSize: 150,
      // preview thumb size , default is 64
      sortDelegate: SortDelegate.common,
      // default is common ,or you make custom delegate to sort your gallery
      checkBoxBuilderDelegate: DefaultCheckBoxBuilderDelegate(
        activeColor: Colors.white,
        unselectedColor: Colors.white,
        checkColor: Colors.green,
      ),
      // default is DefaultCheckBoxBuilderDelegate ,or you make custom delegate to create checkbox

      loadingDelegate: this,
      // if you want to build custom loading widget,extends LoadingDelegate, [see example/lib/main.dart]

      badgeDelegate: const DurationBadgeDelegate(),
      // badgeDelegate to show badge widget

      pickType: type,

      photoPathList: pathList,
    );

    if (imgList == null || imgList.isEmpty) {
      showToast("No pick item.");
      return;
    } else {
      List<String> r = [];
      for (var e in imgList) {
        var file = await e.file;
        r.add(file.absolute.path);
      }
      currentSelected = r.join("\n\n");

      List<AssetEntity> preview = [];
      preview.addAll(imgList);
      Navigator.push(context,
          MaterialPageRoute(builder: (_) => PreviewPage(list: preview)));
    }
    setState(() {});
  }

  void routePage(Widget widget) {
    Navigator.push(
      context,
      MaterialPageRoute(
        builder: (BuildContext context) {
          return widget;
        },
      ),
    );
  }
}
